File: parallel_multiply.cpp
// =============================================================================
//  Parallel Matrix Multiplication in C++
//  Application: Image Convolution Filter (Edge Detection using Sobel Operator)
//
//  Equivalent C++ implementation for direct comparison with the Julia version.
//  Uses OpenMP for multi-threading parallelism.
//
//  Compile:
//    g++ -O3 -fopenmp -std=c++17 -o parallel_multiply parallel_multiply.cpp
//    (MSVC) cl /O2 /openmp /std:c++17 /EHsc parallel_multiply.cpp
// =============================================================================

#include <iostream>
#include <iomanip>
#include <vector>
#include <cmath>
#include <chrono>
#include <random>
#include <algorithm>
#include <numeric>
#include <string>
#include <functional>
#include <cassert>

#ifdef _OPENMP
    #include <omp.h>
#endif

// ─────────────────────────────────────────────────────────────────────────────
//  Matrix Class — Simple 2D Dense Matrix
// ─────────────────────────────────────────────────────────────────────────────
class Matrix {
public:
    int rows, cols;
    std::vector<double> data;

    Matrix() : rows(0), cols(0) {}

    Matrix(int r, int c, double val = 0.0)
        : rows(r), cols(c), data(r * c, val) {}

    inline double& operator()(int i, int j) {
        return data[i * cols + j];
    }

    inline double operator()(int i, int j) const {
        return data[i * cols + j];
    }

    // Fill with random values [0, 1)
    void randomize(unsigned seed = 42) {
        std::mt19937 gen(seed);
        std::uniform_real_distribution<double> dist(0.0, 1.0);
        for (auto& v : data) v = dist(gen);
    }

    // Max absolute difference from another matrix
    double max_diff(const Matrix& other) const {
        assert(rows == other.rows && cols == other.cols);
        double max_err = 0.0;
        for (size_t i = 0; i < data.size(); ++i) {
            max_err = std::max(max_err, std::abs(data[i] - other.data[i]));
        }
        return max_err;
    }
};

// ─────────────────────────────────────────────────────────────────────────────
//  1. NAIVE (SEQUENTIAL) MATRIX MULTIPLICATION
// ─────────────────────────────────────────────────────────────────────────────
/**
 * Standard O(n³) sequential matrix multiplication.
 * Used as a baseline for performance comparison.
 */
Matrix naive_multiply(const Matrix& A, const Matrix& B) {
    int m = A.rows, k = A.cols, n = B.cols;
    Matrix C(m, n);

    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            double s = 0.0;
            for (int p = 0; p < k; ++p) {
                s += A(i, p) * B(p, j);
            }
            C(i, j) = s;
        }
    }
    return C;
}

// ─────────────────────────────────────────────────────────────────────────────
//  2. PARALLEL MATRIX MULTIPLICATION (OpenMP)
// ─────────────────────────────────────────────────────────────────────────────
/**
 * Parallelized matrix multiplication using OpenMP.
 * Distributes row computation across available CPU threads.
 */
Matrix parallel_multiply_openmp(const Matrix& A, const Matrix& B) {
    int m = A.rows, k = A.cols, n = B.cols;
    Matrix C(m, n);

    #pragma omp parallel for schedule(dynamic)
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            double s = 0.0;
            for (int p = 0; p < k; ++p) {
                s += A(i, p) * B(p, j);
            }
            C(i, j) = s;
        }
    }
    return C;
}

// ─────────────────────────────────────────────────────────────────────────────
//  3. BLOCK-PARALLEL MATRIX MULTIPLICATION (Cache-friendly + OpenMP)
// ─────────────────────────────────────────────────────────────────────────────
/**
 * Cache-optimized blocked matrix multiplication with OpenMP.
 * Divides the matrices into sub-blocks for better cache locality.
 */
Matrix parallel_multiply_blocked(const Matrix& A, const Matrix& B, int block_size = 64) {
    int m = A.rows, k = A.cols, n = B.cols;
    Matrix C(m, n);

    #pragma omp parallel for collapse(2) schedule(dynamic)
    for (int bi = 0; bi < m; bi += block_size) {
        for (int bj = 0; bj < n; bj += block_size) {
            int i_end = std::min(bi + block_size, m);
            int j_end = std::min(bj + block_size, n);

            for (int bk = 0; bk < k; bk += block_size) {
                int k_end = std::min(bk + block_size, k);

                for (int i = bi; i < i_end; ++i) {
                    for (int p = bk; p < k_end; ++p) {
                        double a_ip = A(i, p);
                        for (int j = bj; j < j_end; ++j) {
                            C(i, j) += a_ip * B(p, j);
                        }
                    }
                }
            }
        }
    }
    return C;
}

// ─────────────────────────────────────────────────────────────────────────────
//  4. APPLICATION: IMAGE CONVOLUTION (Sobel Edge Detection)
// ─────────────────────────────────────────────────────────────────────────────

// Sobel kernels
constexpr double SOBEL_X[3][3] = {
    {-1,  0,  1},
    {-2,  0,  2},
    {-1,  0,  1}
};

constexpr double SOBEL_Y[3][3] = {
    {-1, -2, -1},
    { 0,  0,  0},
    { 1,  2,  1}
};

/**
 * Generate a synthetic grayscale image with geometric patterns
 * that produce detectable edges.
 */
Matrix generate_synthetic_image(int height, int width) {
    Matrix img(height, width);
    for (int i = 0; i < height; ++i) {
        for (int j = 0; j < width; ++j) {
            if (i > 0.25 * height && i < 0.75 * height &&
                j > 0.25 * width  && j < 0.75 * width) {
                img(i, j) = 200.0;
            } else if (i > 0.1 * height && i < 0.9 * height &&
                       j > 0.1 * width  && j < 0.9 * width) {
                img(i, j) = 100.0;
            } else {
                img(i, j) = static_cast<double>(i + j)
                           / static_cast<double>(height + width) * 50.0;
            }
        }
    }
    return img;
}

/**
 * Apply a 3x3 convolution kernel to an image — SEQUENTIAL.
 */
Matrix apply_convolution_sequential(const Matrix& image, const double kernel[3][3]) {
    int h = image.rows, w = image.cols;
    Matrix output(h, w);

    for (int i = 1; i < h - 1; ++i) {
        for (int j = 1; j < w - 1; ++j) {
            double s = 0.0;
            for (int ki = 0; ki < 3; ++ki) {
                for (int kj = 0; kj < 3; ++kj) {
                    s += image(i - 1 + ki, j - 1 + kj) * kernel[ki][kj];
                }
            }
            output(i, j) = s;
        }
    }
    return output;
}

/**
 * Apply a 3x3 convolution kernel to an image — PARALLEL (OpenMP).
 */
Matrix apply_convolution_parallel(const Matrix& image, const double kernel[3][3]) {
    int h = image.rows, w = image.cols;
    Matrix output(h, w);

    #pragma omp parallel for schedule(dynamic)
    for (int i = 1; i < h - 1; ++i) {
        for (int j = 1; j < w - 1; ++j) {
            double s = 0.0;
            for (int ki = 0; ki < 3; ++ki) {
                for (int kj = 0; kj < 3; ++kj) {
                    s += image(i - 1 + ki, j - 1 + kj) * kernel[ki][kj];
                }
            }
            output(i, j) = s;
        }
    }
    return output;
}

/**
 * Full Sobel edge detection: apply Sobel-X and Sobel-Y, compute gradient magnitude.
 */
Matrix sobel_edge_detect(const Matrix& image, bool parallel = true) {
    Matrix gx, gy;
    if (parallel) {
        gx = apply_convolution_parallel(image, SOBEL_X);
        gy = apply_convolution_parallel(image, SOBEL_Y);
    } else {
        gx = apply_convolution_sequential(image, SOBEL_X);
        gy = apply_convolution_sequential(image, SOBEL_Y);
    }

    Matrix edges(image.rows, image.cols);
    int total = image.rows * image.cols;
    for (int i = 0; i < total; ++i) {
        edges.data[i] = std::sqrt(gx.data[i] * gx.data[i] + gy.data[i] * gy.data[i]);
    }
    return edges;
}

// ─────────────────────────────────────────────────────────────────────────────
//  5. BENCHMARKING UTILITIES
// ─────────────────────────────────────────────────────────────────────────────

using Clock = std::chrono::high_resolution_clock;

/**
 * Benchmark a function with warmup, returning median elapsed time in seconds.
 */
double benchmark(std::function<void()> func, int warmup = 1, int runs = 3) {
    for (int i = 0; i < warmup; ++i) func();

    std::vector<double> times;
    for (int i = 0; i < runs; ++i) {
        auto t0 = Clock::now();
        func();
        auto t1 = Clock::now();
        double elapsed = std::chrono::duration<double>(t1 - t0).count();
        times.push_back(elapsed);
    }
    std::sort(times.begin(), times.end());
    return times[runs / 2]; // median
}

/**
 * Print a formatted line with method name, time, and GFLOP/s.
 */
void print_result(const std::string& name, double time_s, int n) {
    double gflops = (2.0 * n * n * n) / time_s / 1e9;
    std::cout << "    " << std::left << std::setw(25) << name
              << "  " << std::fixed << std::setprecision(4) << std::setw(8) << time_s << " s"
              << "   " << std::setprecision(2) << std::setw(6) << gflops << " GFLOP/s"
              << std::endl;
}

/**
 * Verify correctness of all methods against blocked parallel (as reference).
 */
void verify_correctness(const Matrix& A, const Matrix& B) {
    Matrix ref = parallel_multiply_blocked(A, B);

    struct Method {
        std::string name;
        std::function<Matrix()> func;
    };

    std::vector<Method> methods = {
        {"Naive Sequential",   [&]() { return naive_multiply(A, B); }},
        {"OpenMP Parallel",    [&]() { return parallel_multiply_openmp(A, B); }},
    };

    std::cout << "  Correctness Verification (vs Blocked Parallel):" << std::endl;
    for (auto& m : methods) {
        Matrix result = m.func();
        double max_err = result.max_diff(ref);
        const char* status = (max_err < 1e-10) ? "PASS" : "FAIL";
        std::cout << "    " << std::left << std::setw(25) << m.name
                  << "  max error = " << std::scientific << std::setprecision(2) << max_err
                  << "  " << status << std::endl;
    }
    std::cout << std::endl;
}

// ─────────────────────────────────────────────────────────────────────────────
//  6. MAIN DRIVER
// ─────────────────────────────────────────────────────────────────────────────

int main() {
    std::cout << std::string(72, '=') << std::endl;
    std::cout << "  PARALLEL MATRIX MULTIPLICATION IN C++" << std::endl;
    std::cout << "  Application: Image Processing - Sobel Edge Detection" << std::endl;
    std::cout << std::string(72, '=') << std::endl;
    std::cout << "  C++ Standard  : C++17" << std::endl;
    #ifdef _OPENMP
        std::cout << "  OpenMP Threads: " << omp_get_max_threads() << std::endl;
    #else
        std::cout << "  OpenMP        : NOT AVAILABLE (single-threaded fallback)" << std::endl;
    #endif
    std::cout << std::string(72, '=') << std::endl;

    // ── Part A: Matrix Multiplication Benchmark ─────────────────────────
    std::cout << "\n+-------------------------------------------------------------+" << std::endl;
    std::cout << "|  PART A: Dense Matrix Multiplication Benchmark              |" << std::endl;
    std::cout << "+-------------------------------------------------------------+\n" << std::endl;

    std::vector<int> sizes = {256, 512, 1024};

    for (int n : sizes) {
        std::cout << "  Matrix Size: " << n << " x " << n << std::endl;
        std::cout << "  " << std::string(55, '-') << std::endl;

        Matrix A(n, n), B(n, n);
        A.randomize(42);
        B.randomize(123);

        if (n <= 512) {
            verify_correctness(A, B);
        }

        std::cout << "  Benchmark Results:" << std::endl;

        if (n <= 512) {
            double t_naive = benchmark([&]() { naive_multiply(A, B); });
            print_result("Naive Sequential", t_naive, n);
        }

        double t_omp = benchmark([&]() { parallel_multiply_openmp(A, B); });
        print_result("OpenMP Parallel", t_omp, n);

        double t_blocked = benchmark([&]() { parallel_multiply_blocked(A, B); });
        print_result("Blocked Parallel", t_blocked, n);

        std::cout << std::endl;
    }

    // ── Part B: Image Convolution Application ───────────────────────────
    std::cout << "\n+-------------------------------------------------------------+" << std::endl;
    std::cout << "|  PART B: Image Convolution - Sobel Edge Detection           |" << std::endl;
    std::cout << "+-------------------------------------------------------------+\n" << std::endl;

    std::vector<std::pair<int,int>> img_sizes = {{512, 512}, {1024, 1024}, {2048, 2048}};

    for (auto& [h, w] : img_sizes) {
        std::cout << "  Image Size: " << h << " x " << w << " pixels" << std::endl;
        std::cout << "  " << std::string(45, '-') << std::endl;

        Matrix image = generate_synthetic_image(h, w);

        double t_seq = benchmark([&]() { sobel_edge_detect(image, false); });
        double t_par = benchmark([&]() { sobel_edge_detect(image, true); });

        double speedup = t_seq / t_par;

        Matrix edges = sobel_edge_detect(image, true);
        int edge_pixels = 0;
        for (auto& v : edges.data) {
            if (v > 50.0) ++edge_pixels;
        }

        std::cout << std::fixed;
        std::cout << "    Sequential         : " << std::setprecision(4) << std::setw(8) << t_seq << " s" << std::endl;
        #ifdef _OPENMP
            std::cout << "    Parallel (" << omp_get_max_threads() << " thds) : "
                      << std::setprecision(4) << std::setw(8) << t_par << " s" << std::endl;
        #else
            std::cout << "    Parallel (1 thd)   : " << std::setprecision(4) << std::setw(8) << t_par << " s" << std::endl;
        #endif
        std::cout << "    Speedup            : " << std::setprecision(2) << speedup << "x" << std::endl;
        std::cout << "    Edges detected     : " << edge_pixels << " pixels ("
                  << std::setprecision(1) << (100.0 * edge_pixels / (h * w)) << "%)" << std::endl;
        std::cout << std::endl;
    }

    // ── Summary ─────────────────────────────────────────────────────────
    std::cout << std::string(72, '=') << std::endl;
    std::cout << "  SUMMARY" << std::endl;
    std::cout << std::string(72, '=') << std::endl;
    std::cout << "  C++ requires more boilerplate than Julia for parallel computing:" << std::endl;
    std::cout << "    - Manual memory management with Matrix class" << std::endl;
    std::cout << "    - Explicit OpenMP pragmas for parallelism" << std::endl;
    std::cout << "    - Separate compilation flags needed (-fopenmp)" << std::endl;
    std::cout << "    - No built-in BLAS (requires external library like MKL/OpenBLAS)" << std::endl;
    std::cout << "    - More verbose syntax for the same operations" << std::endl;
    std::cout << std::string(72, '=') << std::endl;

    return 0;
}

File: parallel_multiply.jl
#!/usr/bin/env julia
# =============================================================================
#  Parallel Matrix Multiplication in Julia
#  Application: Image Convolution Filter (Edge Detection using Sobel Operator)
#
#  Demonstrates Julia's native multi-threading and distributed computing
#  capabilities for parallel matrix multiplication — applied to a real-world
#  image processing pipeline.
# =============================================================================

using Base.Threads
using LinearAlgebra
using Printf
using Random
using Statistics

# ─────────────────────────────────────────────────────────────────────────────
#  1. NAIVE (SEQUENTIAL) MATRIX MULTIPLICATION
# ─────────────────────────────────────────────────────────────────────────────
"""
    naive_multiply(A, B)

Standard O(n³) sequential matrix multiplication.
Used as a baseline for performance comparison.
"""
function naive_multiply(A::Matrix{Float64}, B::Matrix{Float64})::Matrix{Float64}
    m, k = size(A)
    _, n = size(B)
    C = zeros(Float64, m, n)

    for i in 1:m
        for j in 1:n
            s = 0.0
            for p in 1:k
                s += A[i, p] * B[p, j]
            end
            C[i, j] = s
        end
    end
    return C
end

# ─────────────────────────────────────────────────────────────────────────────
#  2. PARALLEL MATRIX MULTIPLICATION (Multi-threaded)
# ─────────────────────────────────────────────────────────────────────────────
"""
    parallel_multiply_threads(A, B)

Parallelized matrix multiplication using Julia's `Threads.@threads`.
Distributes row computation across available CPU threads.
"""
function parallel_multiply_threads(A::Matrix{Float64}, B::Matrix{Float64})::Matrix{Float64}
    m, k = size(A)
    _, n = size(B)
    C = zeros(Float64, m, n)

    @threads for i in 1:m
        for j in 1:n
            s = 0.0
            for p in 1:k
                s += A[i, p] * B[p, j]
            end
            C[i, j] = s
        end
    end
    return C
end

# ─────────────────────────────────────────────────────────────────────────────
#  3. BLOCK-PARALLEL MATRIX MULTIPLICATION (Cache-friendly + Threaded)
# ─────────────────────────────────────────────────────────────────────────────
"""
    parallel_multiply_blocked(A, B; block_size=64)

Cache-optimized blocked matrix multiplication with multi-threading.
Divides the matrices into sub-blocks for better cache locality, then
parallelizes across blocks.
"""
function parallel_multiply_blocked(A::Matrix{Float64}, B::Matrix{Float64};
                                    block_size::Int=64)::Matrix{Float64}
    m, k = size(A)
    _, n = size(B)
    C = zeros(Float64, m, n)

    # Calculate number of blocks
    mb = ceil(Int, m / block_size)
    nb = ceil(Int, n / block_size)
    kb = ceil(Int, k / block_size)

    # Create a list of (block_i, block_j) tasks
    tasks = [(bi, bj) for bi in 1:mb for bj in 1:nb]

    @threads for t in 1:length(tasks)
        bi, bj = tasks[t]
        i_start = (bi - 1) * block_size + 1
        i_end   = min(bi * block_size, m)
        j_start = (bj - 1) * block_size + 1
        j_end   = min(bj * block_size, n)

        for bk in 1:kb
            k_start = (bk - 1) * block_size + 1
            k_end   = min(bk * block_size, k)

            @inbounds for i in i_start:i_end
                for p in k_start:k_end
                    a_ip = A[i, p]
                    for j in j_start:j_end
                        C[i, j] += a_ip * B[p, j]
                    end
                end
            end
        end
    end
    return C
end

# ─────────────────────────────────────────────────────────────────────────────
#  4. BLAS-ACCELERATED MULTIPLICATION (Julia's built-in, for reference)
# ─────────────────────────────────────────────────────────────────────────────
"""
    blas_multiply(A, B)

Uses Julia's built-in BLAS-backed `*` operator.
This is the gold standard for dense matrix multiplication performance.
"""
function blas_multiply(A::Matrix{Float64}, B::Matrix{Float64})::Matrix{Float64}
    return A * B
end

# ─────────────────────────────────────────────────────────────────────────────
#  5. APPLICATION: IMAGE CONVOLUTION (Sobel Edge Detection)
# ─────────────────────────────────────────────────────────────────────────────

# Sobel kernels for edge detection
const SOBEL_X = Float64[-1  0  1;
                         -2  0  2;
                         -1  0  1]

const SOBEL_Y = Float64[-1 -2 -1;
                          0  0  0;
                          1  2  1]

"""
    generate_synthetic_image(height, width)

Generate a synthetic grayscale image (matrix) with patterns
to simulate a real image for edge detection demonstration.
"""
function generate_synthetic_image(height::Int, width::Int)::Matrix{Float64}
    img = zeros(Float64, height, width)
    for i in 1:height
        for j in 1:width
            # Create a pattern with edges (rectangles + gradient)
            if 0.25*height < i < 0.75*height && 0.25*width < j < 0.75*width
                img[i, j] = 200.0
            elseif 0.1*height < i < 0.9*height && 0.1*width < j < 0.9*width
                img[i, j] = 100.0
            else
                img[i, j] = Float64(i + j) / Float64(height + width) * 50.0
            end
        end
    end
    return img
end

"""
    apply_convolution_sequential(image, kernel)

Apply a convolution kernel to an image sequentially.
"""
function apply_convolution_sequential(image::Matrix{Float64},
                                       kernel::Matrix{Float64})::Matrix{Float64}
    h, w = size(image)
    kh, kw = size(kernel)
    pad_h = kh ÷ 2
    pad_w = kw ÷ 2
    output = zeros(Float64, h, w)

    for i in (pad_h+1):(h-pad_h)
        for j in (pad_w+1):(w-pad_w)
            s = 0.0
            for ki in 1:kh
                for kj in 1:kw
                    s += image[i - pad_h + ki - 1, j - pad_w + kj - 1] * kernel[ki, kj]
                end
            end
            output[i, j] = s
        end
    end
    return output
end

"""
    apply_convolution_parallel(image, kernel)

Apply a convolution kernel to an image using multi-threaded parallelism.
"""
function apply_convolution_parallel(image::Matrix{Float64},
                                     kernel::Matrix{Float64})::Matrix{Float64}
    h, w = size(image)
    kh, kw = size(kernel)
    pad_h = kh ÷ 2
    pad_w = kw ÷ 2
    output = zeros(Float64, h, w)

    @threads for i in (pad_h+1):(h-pad_h)
        for j in (pad_w+1):(w-pad_w)
            s = 0.0
            for ki in 1:kh
                for kj in 1:kw
                    s += image[i - pad_h + ki - 1, j - pad_w + kj - 1] * kernel[ki, kj]
                end
            end
            output[i, j] = s
        end
    end
    return output
end

"""
    sobel_edge_detect(image; parallel=true)

Full Sobel edge detection: applies Sobel-X and Sobel-Y kernels,
then computes gradient magnitude √(Gx² + Gy²).
"""
function sobel_edge_detect(image::Matrix{Float64}; parallel::Bool=true)::Matrix{Float64}
    if parallel
        gx = apply_convolution_parallel(image, SOBEL_X)
        gy = apply_convolution_parallel(image, SOBEL_Y)
    else
        gx = apply_convolution_sequential(image, SOBEL_X)
        gy = apply_convolution_sequential(image, SOBEL_Y)
    end
    return sqrt.(gx .^ 2 .+ gy .^ 2)
end

# ─────────────────────────────────────────────────────────────────────────────
#  6. BENCHMARKING UTILITIES
# ─────────────────────────────────────────────────────────────────────────────

"""
    benchmark(f, args...; warmup=1, runs=3)

Benchmark a function with warmup runs, returning median elapsed time in seconds.
"""
function benchmark(f::Function, args...; warmup::Int=1, runs::Int=3)::Float64
    # Warmup
    for _ in 1:warmup
        f(args...)
    end

    times = Float64[]
    for _ in 1:runs
        t = @elapsed f(args...)
        push!(times, t)
    end
    return median(times)
end

"""
    verify_correctness(A, B)

Verify that all multiplication methods produce identical results.
"""
function verify_correctness(A::Matrix{Float64}, B::Matrix{Float64})
    ref = blas_multiply(A, B)

    methods = [
        ("Naive Sequential",    naive_multiply),
        ("Threaded Parallel",   parallel_multiply_threads),
        ("Blocked Parallel",    parallel_multiply_blocked),
    ]

    println("  Correctness Verification (vs BLAS):")
    for (name, method) in methods
        result = method(A, B)
        max_err = maximum(abs.(result .- ref))
        status = max_err < 1e-10 ? "✓ PASS" : "✗ FAIL"
        @printf("    %-25s  max error = %.2e  %s\n", name, max_err, status)
    end
    println()
end

# ─────────────────────────────────────────────────────────────────────────────
#  7. MAIN DRIVER
# ─────────────────────────────────────────────────────────────────────────────

function main()
    println("=" ^ 72)
    println("  PARALLEL MATRIX MULTIPLICATION IN JULIA")
    println("  Application: Image Processing — Sobel Edge Detection")
    println("=" ^ 72)
    @printf("  Julia version : %s\n", VERSION)
    @printf("  Threads       : %d\n", nthreads())
    @printf("  BLAS threads  : %d\n", BLAS.get_num_threads())
    println("=" ^ 72)

    # ── Part A: Matrix Multiplication Benchmark ──────────────────────────
    println("\n┌─────────────────────────────────────────────────────────────┐")
    println("│  PART A: Dense Matrix Multiplication Benchmark             │")
    println("└─────────────────────────────────────────────────────────────┘\n")

    sizes = [256, 512, 1024]

    for n in sizes
        println("  Matrix Size: $(n) × $(n)")
        println("  " * "─" ^ 55)

        Random.seed!(42)
        A = rand(Float64, n, n)
        B = rand(Float64, n, n)

        # Skip naive for large sizes (too slow)
        if n <= 512
            verify_correctness(A, B)
        end

        methods = if n <= 512
            [
                ("Naive Sequential",    () -> naive_multiply(A, B)),
                ("Threaded Parallel",   () -> parallel_multiply_threads(A, B)),
                ("Blocked Parallel",    () -> parallel_multiply_blocked(A, B)),
                ("BLAS (built-in)",     () -> blas_multiply(A, B)),
            ]
        else
            [
                ("Threaded Parallel",   () -> parallel_multiply_threads(A, B)),
                ("Blocked Parallel",    () -> parallel_multiply_blocked(A, B)),
                ("BLAS (built-in)",     () -> blas_multiply(A, B)),
            ]
        end

        println("  Benchmark Results:")
        blas_time = 0.0
        for (name, method) in methods
            t = benchmark(method; warmup=1, runs=3)
            if name == "BLAS (built-in)"
                blas_time = t
            end
            gflops = (2.0 * n^3) / t / 1e9
            @printf("    %-25s  %8.4f s   %6.2f GFLOP/s\n", name, t, gflops)
        end

        # Speedup table
        if blas_time > 0
            println("\n  Speedup vs BLAS:")
            for (name, method) in methods
                t = benchmark(method; warmup=0, runs=1)
                speedup = t / blas_time
                @printf("    %-25s  %.2f×\n", name, speedup)
            end
        end
        println()
    end

    # ── Part B: Image Convolution Application ────────────────────────────
    println("\n┌─────────────────────────────────────────────────────────────┐")
    println("│  PART B: Image Convolution — Sobel Edge Detection          │")
    println("└─────────────────────────────────────────────────────────────┘\n")

    img_sizes = [(512, 512), (1024, 1024), (2048, 2048)]

    for (h, w) in img_sizes
        @printf("  Image Size: %d × %d pixels\n", h, w)
        println("  " * "─" ^ 45)

        image = generate_synthetic_image(h, w)

        t_seq = benchmark(sobel_edge_detect, image; warmup=1, runs=3)
        t_par = benchmark(() -> sobel_edge_detect(image; parallel=true); warmup=1, runs=3)

        speedup = t_seq / t_par
        edges = sobel_edge_detect(image; parallel=true)
        edge_pixels = count(x -> x > 50.0, edges)

        @printf("    Sequential         : %8.4f s\n", t_seq)
        @printf("    Parallel (%d thds) : %8.4f s\n", nthreads(), t_par)
        @printf("    Speedup            : %5.2f×\n", speedup)
        @printf("    Edges detected     : %d pixels (%.1f%%)\n",
                edge_pixels, 100.0 * edge_pixels / (h * w))
        println()
    end

    # ── Summary ──────────────────────────────────────────────────────────
    println("=" ^ 72)
    println("  SUMMARY")
    println("=" ^ 72)
    println("  Julia's key advantages for parallel matrix operations:")
    println("    • Native multi-threading with @threads (zero boilerplate)")
    println("    • Built-in BLAS integration for production workloads")
    println("    • Type-stable code compiles to efficient native code")
    println("    • Easy transition from sequential to parallel")
    println("    • First-class support for linear algebra operations")
    println("=" ^ 72)
end

# Run
main()
